azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java

```diff
@@ -934,7 +939,9 @@ public void cancelExecutingJobs(ExecutableFlow exFlow, String userId,
   @Override
   public String submitExecutableFlow(ExecutableFlow exflow, String userId)
     throws ExecutorManagerException {
-    synchronized (exflow) {
+
+    String exFlowKey = exflow.getProjectName() + "." + exflow.getId() + ".submitFlow";
+    synchronized (exFlowKey.intern()) {
```

Comment:
i think it will fix the race condition when  same flow is submitted by API and schedule at the same time.